sdk/src/main/java/com/mapbox/maps/ResourceOptionsManager.kt

```diff
@@ -24,20 +24,41 @@ class ResourceOptionsManager(
    * Static variables and methods.
    */
   companion object {
-    private lateinit var default: ResourceOptionsManager
+    internal var default: ResourceOptionsManager? = null
+    internal fun getTokenResId(context: Context): Int = context.resources.getIdentifier(
+      MAPBOX_ACCESS_TOKEN_RESOURCE_NAME,
+      "string",
+      context.packageName
+    )
+
+    fun reset() {
+      default = null
+    }
 
     /**
      * The default shared instance with default resource options.
+     *
+     * @param context the application context
+     * @defaultToken the default token will be set to the default [ResourceOptionsManager] instance.
+     * If set to null, will reuse the previously set one or search for an access token in the application resources for the first time of init.
+     * It will throw [MapboxConfigurationException] if there's no token found.
      */
     @Synchronized
-    fun getDefault(context: Context): ResourceOptionsManager {
-      if (!this::default.isInitialized) {
-        default = ResourceOptionsManager(
-          ResourceOptions.Builder().applyDefaultParams(context)
-            .build()
-        )
+    fun getDefault(context: Context, defaultToken: String? = null): ResourceOptionsManager {
+      if (default == null) {
+        val builder = ResourceOptions.Builder().applyDefaultParams(context)
+
+        if (defaultToken != null) {
+          // Apply the user setting token as the default token
+          builder.accessToken(defaultToken)
+        } else {
+          val tokenResId = getTokenResId(context)
+          // Throw exception as no token could be found as default token.
+          if (tokenResId == 0) throw MapboxConfigurationException()
+        }
+        default = ResourceOptionsManager(builder.build())
       }
-      return default
+      return default!!
     }
```

Comment:
We could do something like 
```
    fun getDefault(context: Context, defaultToken: String? = null): ResourceOptionsManager {
      return (default ?: run {
        val builder = ResourceOptions.Builder().applyDefaultParams(context)

        if (defaultToken != null) {
          // Apply the user setting token as the default token
          builder.accessToken(defaultToken)
        } else {
          val tokenResId = getTokenResId(context)
          // Throw exception as no token could be found as default token.
          if (tokenResId == 0) throw MapboxConfigurationException()
        }
        ResourceOptionsManager(builder.build()).also {
          default = it
        }
      }).apply {
        // Always set the provided access token to default
        defaultToken?.let { token ->
          update { accessToken(token) }
        }
      }
    }
  }
```

Reply:
Thanks for the suggestion, have changed to a more clear logic.